File: src/pypaginate/search/parser.py
"""Token parser for search queries.

Splits search queries into tokens, respecting quoted phrases.
No external dependencies -- uses stdlib only.
"""

from __future__ import annotations

import shlex


class TokenParser:
    """Parse search queries into individual tokens.

    Handles quoted phrases as single tokens and splits
    unquoted text on whitespace.
    """

    __slots__ = ()

    def parse(self, query: str) -> list[str]:
        """Split a query string into search tokens.

        Quoted phrases are preserved as single tokens.
        Extra whitespace is stripped.

        Args:
            query: Raw search query (e.g. ``'"john doe" admin'``).

        Returns:
            List of token strings.

        Examples:
            >>> TokenParser().parse('"john doe" admin')
            ['john doe', 'admin']
            >>> TokenParser().parse("  hello   world  ")
            ['hello', 'world']
        """
        stripped = query.strip()
        if not stripped:
            return []
        return self._tokenize(stripped)

    @staticmethod
    def _tokenize(text: str) -> list[str]:
        """Tokenize text using shlex for quote handling.

        Falls back to simple whitespace split if shlex
        encounters malformed quotes.

        Args:
            text: Non-empty, stripped query text.

        Returns:
            List of non-empty token strings.
        """
        try:
            tokens = shlex.split(text)
        except ValueError:
            tokens = text.split()
        return [t for t in tokens if t]


__all__ = ["TokenParser"]

File: crates/core/src/search/parser.rs
//! Quote-aware query tokenizer, ported from pypaginate's `search/parser.py`.
//!
//! Python uses `shlex.split`, falling back to a plain whitespace split when the
//! quotes are unbalanced (`except ValueError`). The `shlex` crate mirrors POSIX
//! shell word-splitting and returns `None` on a parse error, which we map to the
//! same whitespace fallback.

/// Split a query into tokens, honouring quoted phrases.
#[must_use]
pub(crate) fn tokenize(query: &str) -> Vec<String> {
    let stripped = query.trim();
    if stripped.is_empty() {
        return Vec::new();
    }
    match shlex::split(stripped) {
        Some(tokens) => tokens.into_iter().filter(|t| !t.is_empty()).collect(),
        None => stripped.split_whitespace().map(str::to_owned).collect(),
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn quoted_phrases_stay_together() {
        assert_eq!(tokenize(r#""john doe" admin"#), vec!["john doe", "admin"]);
    }

    #[test]
    fn whitespace_collapses() {
        assert_eq!(tokenize("  hello   world  "), vec!["hello", "world"]);
    }

    #[test]
    fn empty_query_has_no_tokens() {
        assert!(tokenize("   ").is_empty());
    }

    #[test]
    fn unbalanced_quotes_fall_back_to_whitespace() {
        // shlex rejects the dangling quote; fallback splits on whitespace.
        assert_eq!(tokenize(r#"foo "bar"#), vec!["foo", r#""bar"#]);
    }
}
